### PR TITLE
ipmi: Retry sending key in grub

### PR DESCRIPTION
**Pull request by Soulofdestiny**
[progress ticket](https://progress.opensuse.org/issues/45362)

### CODE REVIEW

File: tests/installation/grub_test.pm [hunk 1]
```diff
@@ -100,7 +102,7 @@ sub run {
     }
     else {
         # avoid timeout for booting to HDD
-        send_key 'ret';
+        send_key 'ret' for (1 .. $retry);
```
Comment (coolo) [inline tests/installation/grub_test.pm]: I wouldn't call this $retry - it's just a loop without any expectations in between. And if the key is not lost, you just added 4 additional enter presses into the keyboard buffer. You will have fun finding the fallout of those :)

Comment (Soulofdestiny) [reply tests/installation/grub_test.pm]: damn, i thought no one will notice ;)

Comment (coolo) [reply tests/installation/grub_test.pm]: As okurz keeps a record of my quotes - ask him for the Pappenheimer :)

**Comment by okurz** [comment]
Hm, When we do this in grub_test would we not need this in `wait_boot` as well? If yes (what I think), then the proper solution of course is not to copy-paste it. What if the key press is also lost elsewhere?
Also if it annoys I would also go for a multi-tag assert_screen here and not check for 10 seconds if we are still in the grub menu. Who knows, it might still be visible on the screen after 10 seconds, you send a second key press and you end up with what coolo described. Please also see my comment in the ticket.

File: tests/installation/grub_test.pm [hunk 2]
```diff
@@ -101,6 +103,10 @@ sub run {
     else {
         # avoid timeout for booting to HDD
         send_key 'ret';
+        if ($retry) {
+            # on retry, wait 10 seconds if grub2 is still shown, if yes, send the key again
+            check_screen('grub2', 10) ? return : send_key 'ret';
```
Comment (kravciak) [inline tests/installation/grub_test.pm]: Aren't you doing the opposite? - You send 'ret' when grub2 is not visible.

File: tests/installation/grub_test.pm [hunk 3]
```diff
@@ -81,6 +81,8 @@ sub bug_workaround_bsc1005313 {
 sub run {
     my ($self) = @_;
     my $timeout = get_var('GRUB_TIMEOUT', 90);
+    # on IPMI keys are sometimes lost in the bootloader
+    my $retry = check_var('BACKEND','ipmi') ? 1 : undef);
```
Comment (kravciak) [inline tests/installation/grub_test.pm]: extra `)`
You might not even need `$retry` var since it's used only once - but whatever is more readable for you.

**Comment by okurz** [comment]
Well, let's try, triggered for verification:

```
bash <(curl -s https://raw.githubusercontent.com/okurz/scripts/master/openqa-clone-custom-git-refspec) \
  https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/6646 \
  https://openqa.suse.de/tests/2401679
```

Created job #2405825: sle-15-SP1-Installer-DVD-x86_64-Build149.2-btrfs@64bit-ipmi -> https://openqa.suse.de/t2405825

as expected, incompletes, see https://openqa.suse.de/tests/2405825/file/autoinst-log.txt with

```
[2019-01-26T07:57:34.109 CET] [debug] error on tests/installation/grub_test.pm: syntax error at /var/lib/openqa/pool/23/os-autoinst-distri-opensuse/tests/installation/grub_test.pm line 85, near "undef)"
Compilation failed in require at (eval 1434) line 1.

error on tests/installation/grub_test.pm: syntax error at /var/lib/openqa/pool/23/os-autoinst-distri-opensuse/tests/installation/grub_test.pm line 85, near "undef)"
```

**Pull request closed without merging**